src/services/notification.service.js

```diff
@@ -0,0 +1,71 @@
+"use strict";
+const notificationModel = require("../models/notification.model");
+
+const pushNotificationToSystem = async ({
+  type = "SHOP-001",
+  receivedId,
+  senderId,
+  options,
+}) => {
+  let noti_content;
+  if (type === "SHOP-001") {
+    noti_content = "@@@ vua moi them mot san pham";
+  } else if (type === "PROMOTION-001") {
+    noti_content = "@@@ vua moi them mot voucher";
+  }
+
+  const newNoti = await notificationModel.create({
+    noti_type: type,
+    noti_content,
+    noti_senderId: senderId,
+    noti_receiveId: receivedId,
+    noti_options: options,
+  });
+
+  return newNoti;
+};
+
+const listNotificationByUser = async ({
+  userId = 1,
+  type = "All",
+  isRead = 0,
+}) => {
+  const match = {
+    noti_receiveId: userId,
+  };
+  if (type !== "All") {
+    match["noti_type"] = type;
+  }
+
+  return await notificationModel.aggregate([
```

Comment:
Hàm return await notificationModel.aggregate([...]) sử dụng phương thức aggregate của MongoDB để thực hiện một truy vấn pipeline trên tập hợp (collection) notificationModel. Dưới đây là chi tiết giải thích từng bước của pipeline được sử dụng:

1. Pipeline tổng quan
Hàm này trả về một danh sách thông báo đã được xử lý (aggragate), với dữ liệu tùy chỉnh theo các yêu cầu như gộp các trường, thay đổi nội dung trường noti_content, và lấy các trường cụ thể từ mỗi tài liệu trong tập hợp notificationModel.

2. Bước $match
javascript
Sao chép mã
{
  $match: match,
}
Mục đích: Lọc các tài liệu (document) trong tập hợp notificationModel sao cho chúng thỏa mãn điều kiện match.
match là một biến đầu vào, có thể là một object chứa các tiêu chí lọc (ví dụ: { noti_type: 'new_product' }).
Tương tự như câu lệnh WHERE trong SQL.
3. Bước $project
javascript
Sao chép mã
{
  $project: {
    noti_type: 1,
    noti_senderId: 1,
    noti_receiveId: 1,
    noti_content: 1,
    createAt: 1,
    noti_options: 1,
    noti_content: {
      $concat: [
        {
          $substr: ["$noti_options.shop_name", 0, -1],
        },
        "vua them mot san pham",
        {
          $substr: ["$noti_options.product_name", 0, -1],
        },
      ],
    },
  },
}
Mục đích: Chỉ giữ lại các trường cần thiết trong kết quả và tạo trường noti_content mới với nội dung được ghép từ các giá trị khác.
Các trường gốc như noti_type, noti_senderId, noti_receiveId, createAt, và noti_options được giữ lại (1 nghĩa là bao gồm trường này).
Tạo mới noti_content:
Sử dụng $concat để kết hợp chuỗi:
Phần đầu tiên là giá trị shop_name từ noti_options, được cắt chuỗi bằng $substr.
Tiếp theo là chuỗi tĩnh "vua them mot san pham".
Cuối cùng là giá trị product_name từ noti_options, cũng được cắt chuỗi bằng $substr.
Cú pháp $substr: ["<trường>", <vị trí bắt đầu>, <độ dài>].
Ở đây, độ dài -1 có nghĩa là lấy toàn bộ chuỗi (đến hết chuỗi).
4. Kết quả
Kết quả trả về từ phương thức aggregate sẽ là một mảng các tài liệu với định dạng tương tự sau khi qua pipeline:

json
Sao chép mã
[
  {
    "noti_type": "new_product",
    "noti_senderId": "12345",
    "noti_receiveId": "67890",
    "noti_content": "Shop ABC vua them mot san pham Product XYZ",
    "createAt": "2024-12-15T12:00:00Z",
    "noti_options": {
      "shop_name": "Shop ABC",
      "product_name": "Product XYZ"
    }
  },
  ...
]
5. Cách sử dụng trong mô hình (model) MongoDB
Hàm trên thường được đặt trong một file model của MongoDB. Khi được gọi, nó sẽ:

Lọc dữ liệu thông báo dựa trên điều kiện trong match.
Biến đổi dữ liệu để tạo trường noti_content mới.
Trả về danh sách thông báo đã được xử lý.
6. Điểm chú ý
Hiệu suất: Nếu dữ liệu lớn, cần đảm bảo rằng các trường được lọc (match) đã được đánh chỉ mục (indexed) để tăng tốc độ truy vấn.
Xử lý chuỗi: Sử dụng $substr với độ dài -1 phù hợp khi không biết trước độ dài chuỗi. Tuy nhiên, nếu chuỗi có thể chứa giá trị null, cần thêm kiểm tra để tránh lỗi.